### PR TITLE
[SwiftSyntax] Add SyntaxRewriter.visitAny(_:)

### PR DESCRIPTION
**Pull request by harlanhaskins**
This patch makes it so clients can override `visit(_ node: Syntax) -> Syntax` in `SyntaxRewriter` and `SyntaxVisitor` subclasses, if they choose to override the global visitation behavior.

### CODE REVIEW

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by nkcsgexi** [comment]
hmm, this will make `visitPre` and `visitPost` weird after clients rewriting `visit()`. Can we add another open function instead of making this one open?

**Comment by harlanhaskins** [comment]
Ah, yes you're right - this would make it so `visitPost` didn't happen after `visit(_:)`.

Something like `open func visitAny(_ syntax: Syntax) -> Syntax`?

**Comment by nkcsgexi** [comment]
sounds good!

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

File: tools/SwiftSyntax/SyntaxRewriter.swift.gyb [hunk 1]
```diff
@@ -45,21 +45,31 @@ open class SyntaxRewriter {
...
 
   public func visit(_ node: Syntax) -> Syntax {
     visitPre(node)
     defer { visitPost(node) }
-    switch node.raw.kind {
-    case .token: return visit(node as! TokenSyntax)
+
+    // Visit this node using the global visitor first
+    let newNode = visitAny(node)
+    switch newNode.raw.kind {
```
[the remark this answers is not in the document]
Comment (harlanhaskins) [reply tools/SwiftSyntax/SyntaxRewriter.swift.gyb]: That makes sense, I'll do that!

File: tools/SwiftSyntax/SyntaxRewriter.swift.gyb [hunk 2]
```diff
@@ -45,13 +45,29 @@ open class SyntaxRewriter {
...
   open func visitPost(_ node: Syntax) {}
 
   public func visit(_ node: Syntax) -> Syntax {
     visitPre(node)
     defer { visitPost(node) }
+
+    // If the global visitor returned non-nil, visit the rewritten node's
+    // children and skip specialized dispatch.
+    if let newNode = visitAny(node) {
+      return visitChildren(newNode)
```
Comment (nkcsgexi) [inline tools/SwiftSyntax/SyntaxRewriter.swift.gyb]: hmm, sorry about being a little nit-picky here. Whether `visitingChildren` should be up to the overriding function, right? like other specialized visit functions.

Comment (harlanhaskins) [reply tools/SwiftSyntax/SyntaxRewriter.swift.gyb]: Oh, yes you're right. Thanks for catching that!

Comment (harlanhaskins) [reply tools/SwiftSyntax/SyntaxRewriter.swift.gyb]: The more I think about it, though, the more I wonder if we _should_ always call visitChildren after the specialized visitor has been called.

For example, if I'm writing a transform that, say, lays out function call arguments a certain way, I shouldn't have to manually iterate over all subexpressions and visit them, no?

Comment (nkcsgexi) [reply tools/SwiftSyntax/SyntaxRewriter.swift.gyb]: yeah, i think it is a valid assumption that in most cases users will visit a node's children even though the node itself has been updated. @rintaro what do you think?

Comment (rintaro) [reply tools/SwiftSyntax/SyntaxRewriter.swift.gyb]: Perhaps, we should separate `Visitor` and `Walker` like `ASTVisitor` / `ASTWalker`.

**Review by nkcsgexi — APPROVED** [review]

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test and merge

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Comment by harlanhaskins** [comment]
@swift-ci please smoke test

**Pull request merged**